README.md
# Ethereum Smart Contract for the WAXP ERC20 tokens

An `ERC20` standard token with burn capability. Built on the [OpenZeppelin](https://openzeppelin.org/) framework.

Specifically, this contract is upgradeable, relying on the upgradeable official fork of openzeppelin. See: https://github.com/OpenZeppelin/openzeppelin-contracts-ethereum-package#using-via-the-openzeppelin-cli

Only contract owner can burn token.

### Deployments

Production token address: 

Rinkeby token address: *0x0de41b96376f7AaFf671BE23c7881c695a0d161D

### Setup

Be sure to have node >= 10.x

```
$ node --version
v10.16.3
```

Install dev dependencies:

```
$ npm install
```

### Test

```
$ npm run test
```

### Test coverage

```
$ npm run coverage

----------------|----------|----------|----------|----------|----------------|
File            |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
----------------|----------|----------|----------|----------|----------------|
 contracts/     |      100 |      100 |      100 |      100 |                |
  WAXPERC20.sol |      100 |      100 |      100 |      100 |                |
----------------|----------|----------|----------|----------|----------------|
All files       |      100 |      100 |      100 |      100 |                |
----------------|----------|----------|----------|----------|----------------|
```

### Upgrade

You will have to transfer ownership back to a 12-word mnemonic (`npx mnemonics`) controlled address rather than a ledger device as the support for it with deployments is poor. Use the `PROJECT_ID=<infura web3 provider project id> MNEMONIC="<12 word mnemonic> npm run change-owner` command to do so and change fields appropriately in the scripts/owner file to hit the correct new ownership addresses. Realize the proxyAdmin accounts within the owner.js file are correct for rinkeby and mainnent. Once the proxyAdmin wonership is trnasferred to a mnemonic controlled address, use it to upgrade the contract via `PROJECT_ID=<infura web3 provider project id> MNEMONIC="<12 word mnemonic> npm run upgrade`. After the upgrade is complete, return control back to the ledger using the change onwner command. You will have to edit the scripts/owner.js appropriately to do so.

### Deploy and Info About Upgradeable Contracts

See:
* https://docs.openzeppelin.com/learn/connecting-to-public-test-networks
* https://docs.openzeppelin.com/learn/preparing-for-mainnet


Note: see this GH thread to better understand the ownership vs admin roles in the upgradable contracts pattern https://github.com/OpenZeppelin/openzeppelin-sdk/issues/968#issuecomment-501899483

#### Upgrades contracts summary
* The ProxyAdmin is a contract which is by default the admin (i.e., the one that can upgrade Proxy instances or change a Proxy instance admin) of your created instances.
* The Proxy is a contract that delegates calls to a logic contract, probably the one you developed. The proxy address is the one you provide to your users.
* The Owner in this context it is the account you have used to create a proxy (i.e., the one specified with the --from flag or a default account), who is the owner of the ProxyAdmin (using openzeppelin-solidity Ownable contract), but not the admin of the Proxy (because, as said before, the default admin of the Proxy will be the ProxyAdmin.
* This mean that when you use the zos set-admin command, you will be changing the admin of a Proxy, from the ProxyAdmin to any other account you provide. This is really dangerous and you should basically never do this. No idea why this is a useful command. Whereas changing the owner of the proxyadmin contract is useful because that user controls upgrades so you would want to secure that with an offline account

### Truffle deploy and upgrade

#### Migrate from OpenZeppelin CLI to truffle

Reference: https://docs.openzeppelin.com/upgrades-plugins/1.x/migrate-from-cli

#### Deploy new upgradeable contract

- create deployment script `migrations/1_deploy.js`

```javascript
const { deployProxy } = require('@openzeppelin/truffle-upgrades');

const WAXPERC20UpgradeSafe = artifacts.require('WAXPERC20UpgradeSafe');

module.exports = async function (deployer) {
  const instance = await deployProxy(WAXPERC20UpgradeSafe, [process.env.ESCROW_ADDRESS], { deployer });
  console.log('Deployed', instance.address);
};
```

- run truffle migration:

```bash
MNEMONIC="YOUR MNEMONIC" PROJECT_ID="YOUR PROJECT ID" npx truffle migrate --network rinkeby
```

#### Upgrade contract

- create upgrade script `migrations/2_upgrade.js`

```javascript
const { upgradeProxy } = require('@openzeppelin/truffle-upgrades');

const WAXPERC20UpgradeSafe = artifacts.require('WAXPERC20UpgradeSafe');
const WAXPERC20UpgradeSafeV2 = artifacts.require('WAXPERC20UpgradeSafeV2');

module.exports = async function (deployer) {
  const existing = await WAXPERC20UpgradeSafe.deployed();
  const instance = await upgradeProxy(existing.address, WAXPERC20UpgradeSafeV2, { deployer });
  console.log("Upgraded", instance.address);
};
```

- run truffle migration:

```bash
MNEMONIC="YOUR MNEMONIC" PROJECT_ID="YOUR PROJECT ID" npx truffle migrate --network rinkeby
```

#### Interact with contract

```bash
MNEMONIC="YOUR MNEMONIC" PROJECT_ID="YOUR PROJECT ID" npx truffle console --network rinkeby
truffle(rinkeby)> let instance = await WAXPERC20UpgradeSafe.deployed()
truffle(rinkeby)> (await instance.totalSupply()).toString()
'386482894311326596'
truffle(rinkeby)> await instance.transfer("0x26e7ef2d05793c6d47c678f1f4b246856236f089", "400000000");
truffle(rinkeby)> (await instance.balanceOf("0x26e7ef2d05793c6d47c678f1f4b246856236f089")).toString()
'400000000'
```

#### setup swapogwax

1. set old wax token address

```bash
MNEMONIC="YOUR MNEMONIC" PROJECT_ID="YOUR PROJECT ID" npx truffle console --network rinkeby
truffle(rinkeby)> let instance = await WAXPERC20UpgradeSafe.deployed()
truffle(rinkeby)> await instance.setWaxToken("0x3Fc977975c70ffc9d2C2A6aCb3Ea43EEF50d21B3")
```

2. Transfer amount of WAXP to contract address

```bash
await instance.transfer("0x1DA0e4b6609eE707379b9f65225c7589B3c8aB94", "1000000000000000")
```

3. User approve for new WAXP contract to take token in old WAX contract

```bash
truffle(rinkeby)> const oldWaxToken = await WAXPERC20UpgradeSafe.at("0x3Fc977975c70ffc9d2C2A6aCb3Ea43EEF50d21B3")
truffle(rinkeby)> await oldWaxToken.approve("0x1DA0e4b6609eE707379b9f65225c7589B3c8aB94", "100000000")
```

4. Swap WAXP from old contract to new contract

```bash
truffle(rinkeby)> await instance.swapogwax("1000000000")
```

CHANGELOG.md
# Changelog

## v1.1.0

BREAKING CHANGES:

FEATURES:
- Add the ability to swap 1-1 OG WAX for WAXP

IMPROVEMENTS:

BUG FIXES:

## v1.0.0

BREAKING CHANGES:

FEATURES:
- Burnable WAXP ERC20 Fungible Token

IMPROVEMENTS:

BUG FIXES:


test/test-token.js
// Create a contract object from a compilation artifact
const truffleAssert = require('truffle-assertions');
const WAXPERC20 = artifacts.require("WAXPERC20UpgradeSafe");

contract('WAXPERC20', accounts => {
  const [ owner, escrow, user ] = accounts;
  const initialTotalSupply = web3.utils.toBN('386482894311326596');
  let waxpERC20contract;
  let oldWaxTokenContract;

  beforeEach(async function () {
    this.timeout(10000);
    // Deploy a new WAXPERC20 contract for each test
    waxpERC20contract = await WAXPERC20.new();
    oldWaxTokenContract = await WAXPERC20.new(); // simulate old WAX ERC20 contract
    await waxpERC20contract.initialize(escrow, { from: owner });
    await oldWaxTokenContract.initialize(escrow, { from: owner });
    await waxpERC20contract.setWaxToken(oldWaxTokenContract.address, { from: owner});
  });

  it('initializes', async function () {
    // Test if the supply is as expected
    expect((await waxpERC20contract.totalSupply()).eq(initialTotalSupply)).to.be.true;
    expect((await waxpERC20contract.balanceOf(escrow)).eq(initialTotalSupply)).to.be.true;
    expect((await waxpERC20contract.symbol())).to.equal('WAXP');
    expect((await waxpERC20contract.name())).to.equal('WAXP Token');
    expect((await waxpERC20contract.decimals()).eq(web3.utils.toBN('8'))).to.be.true;

    await truffleAssert.reverts(
      waxpERC20contract.initialize(escrow, { from: escrow }),
      'contract is already initialized'
    );
  });

  it('transfers', async function () {
    const amount = web3.utils.toBN('100');
    await waxpERC20contract.transfer(user, amount, { from: escrow });
    expect((await waxpERC20contract.balanceOf(escrow)).eq(initialTotalSupply.sub(amount))).to.be.true;
    expect((await waxpERC20contract.balanceOf(user)).eq(amount)).to.be.true;
    await truffleAssert.reverts(
      waxpERC20contract.transfer(user, initialTotalSupply, { from: escrow }),
      'transfer amount exceeds balance'
    );
  });

  it('should not allow transfer() when to is 0x0000000000000000000000000000000000000000', async function() {
    const amount = web3.utils.toBN('100');
    await truffleAssert.reverts(
      waxpERC20contract.transfer('0x0000000000000000000000000000000000000000', amount, { from: escrow }),
      'transfer to the zero address'
    );
  })

  it('should allow transferFrom(), when properly approved, when unpaused', async function() {
    const amount = web3.utils.toBN('100');
    let receipt = await waxpERC20contract.approve(user, amount, { from: escrow });
    truffleAssert.eventEmitted(receipt, 'Approval', (ev) => {
      return ev.value.eq(amount)
    }, 'Invalid approval event');
    await waxpERC20contract.transferFrom(escrow, user, amount, { from: user });
    expect((await waxpERC20contract.balanceOf(escrow)).eq(initialTotalSupply.sub(amount))).to.be.true;
    expect((await waxpERC20contract.balanceOf(user)).eq(amount)).to.be.true;
  })

  it('should allow approve(), and allowance() when unpaused', async function() {
    const amount = web3.utils.toBN('100');
    let receipt = await waxpERC20contract.approve(user, amount, { from: escrow });
    truffleAssert.eventEmitted(receipt, 'Approval', (ev) => {
      return ev.value.eq(amount)
    }, 'Invalid approval event');
    expect((await waxpERC20contract.allowance(escrow, user, { from: escrow })).eq(amount)).to.be.true;
    await waxpERC20contract.approve(user, 0, { from: escrow });
  })

  it('should not allow transferFrom() when to is 0x0000000000000000000000000000000000000000', async function() {
    const amount = web3.utils.toBN('100');
    await waxpERC20contract.approve(user, amount, { from: escrow });
    await truffleAssert.reverts(
      waxpERC20contract.transferFrom(escrow, '0x0000000000000000000000000000000000000000', amount, { from: user }),
      'transfer to the zero address'
    );
    await waxpERC20contract.approve(user, 0, { from: escrow });
  })

  it('should not allow burn token if not contract owner', async function() {
    const amount = web3.utils.toBN('100');
    await waxpERC20contract.transfer(user, amount, { from: escrow });

    await truffleAssert.reverts(
      waxpERC20contract.burn(amount, { from: user }),
      'Ownable: caller is not the owner'
    );

    const userBalance = await waxpERC20contract.balanceOf(user);
    expect(userBalance.eq(amount)).to.be.true;
  });

  it('should allow contract owner to burn token', async function() {
    const amount = web3.utils.toBN('10000000000000');
    await waxpERC20contract.transfer(owner, amount, { from: escrow });

    await waxpERC20contract.burn(amount, { from: owner });

    const totalSupply = await waxpERC20contract.totalSupply();
    const userBalance = await waxpERC20contract.balanceOf(owner);
    expect(userBalance.eq(web3.utils.toBN('0'))).to.be.true;
    expect(totalSupply.eq(web3.utils.toBN('386472894311326596'))).to.be.true;
  });

  it('should not allow to burnFrom if not contract owner', async function() {
    const burnAmount = web3.utils.toBN('10000000000000');
    await waxpERC20contract.approve(user, burnAmount, { from: escrow });
    await truffleAssert.reverts(
      waxpERC20contract.burnFrom(escrow, burnAmount, { from: user }),
      'Ownable: caller is not the owner'
    );

    const userAllowance = await waxpERC20contract.allowance(escrow, user);
    expect(userAllowance.eq(burnAmount)).to.be.true;
  });

  it('should allow contract owner to burnFrom', async function() {
    const burnAmount = web3.utils.toBN('10000000000000');
    await waxpERC20contract.approve(owner, burnAmount, { from: escrow });

    await waxpERC20contract.burnFrom(escrow, burnAmount, { from: owner });
    const totalSupply = await waxpERC20contract.totalSupply();
    const userAllowance = await waxpERC20contract.allowance(escrow, owner);
    expect(totalSupply.eq(web3.utils.toBN('386472894311326596'))).to.be.true;
    expect(userAllowance.toString()).to.equal('0');
  });

  it('should not allow swapogwax if not enough allowance', async function() {
    const swapAmount = web3.utils.toBN('10000000000000');
    await oldWaxTokenContract.transfer(user, swapAmount, { from: escrow });

    await truffleAssert.reverts(
      waxpERC20contract.swapogwax(swapAmount, { from: user }),
      'ERR::NOT_ENOUGH_ALLOWANCE'
    )

    await waxpERC20contract.approve(owner, swapAmount.sub(web3.utils.toBN('100000000000')), { from: escrow });

    await truffleAssert.reverts( // still not enough allowance
      waxpERC20contract.swapogwax(swapAmount, { from: user }),
      'ERR::NOT_ENOUGH_ALLOWANCE'
    )
  });

  it('should able to swapogwax to get new waxp token', async function() {
    const swapAmount = web3.utils.toBN('10000000000000');
    await waxpERC20contract.transfer(waxpERC20contract.address, swapAmount, { from: escrow });
    await oldWaxTokenContract.transfer(user, swapAmount, { from: escrow });

    await oldWaxTokenContract.approve(waxpERC20contract.address, swapAmount, { from: user });

    const userBalanceBeforeSwap = await waxpERC20contract.balanceOf(user);
    const receipt = await waxpERC20contract.swapogwax(swapAmount, { from: user });
    const userBalanceAfterSwap = await waxpERC20contract.balanceOf(user);
    expect(userBalanceBeforeSwap.add(swapAmount).eq(userBalanceAfterSwap)).to.be.true;

    truffleAssert.eventEmitted(receipt, 'TokenSwap', (ev) => {
      return ev.from === user && ev.amount.eq(swapAmount)
    }, 'Invalid TokenSwap event');
  });
});

networks.js
const HDWalletProvider = require('@truffle/hdwallet-provider');
const assert = require('assert');

function requireEnvVar(envVar, description) {
  assert(process.env[envVar] !== undefined, `Env var ${envVar} must be present${description ? ' - ' + description : ''}`);
  return process.env[envVar];
}

module.exports = {
  networks: {
    development: {
      protocol: 'http',
      host: process.env['ETHEREUM_HOST'] || 'localhost',
      port: 8545,
      gas: 5000000,
      gasPrice: 5e9,
      networkId: '*',
    },
    rinkeby: {
      provider: () => {
        require('dotenv').config({path: '.env.rinkeby'});
        const projectId = requireEnvVar('PROJECT_ID', 'Infura project id');
        const mnemonic = requireEnvVar('MNEMONIC', 'HD Wallet mnemonic for deployment');
        return new HDWalletProvider(
          mnemonic, `https://rinkeby.infura.io/v3/${projectId}`
        )
      },
      networkId: 4,
      gasPrice: 10e9
    },
    ropsten: {
      provider: () => {
        require('dotenv').config({path: '.env.ropsten'});
        const projectId = requireEnvVar('PROJECT_ID', 'Infura project id');
        const mnemonic = requireEnvVar('MNEMONIC', 'HD Wallet mnemonic for deployment');
        return new HDWalletProvider(
          mnemonic, `https://ropsten.infura.io/v3/${projectId}`
        )
      },
      networkId: 3,
      gasPrice: 10e9
    },
    mainnet: {
      provider: () => {
        require('dotenv').config({path: '.env.mainnet'});
        const projectId = requireEnvVar('PROJECT_ID', 'Infura project id');
        const mnemonic = requireEnvVar('MNEMONIC', 'HD Wallet mnemonic for deployment');
        return new HDWalletProvider(
          mnemonic, `https://mainnet.infura.io/v3/${projectId}`
        )
      },
      networkId: 1,
      gasPrice: 42e9
    },
    ropsten: {
      provider: () => {
        require('dotenv').config({path: '.env.ropsten'});
        const projectId = requireEnvVar('PROJECT_ID', 'Infura project id');
        const mnemonic = requireEnvVar('MNEMONIC', 'HD Wallet mnemonic for deployment');
        return new HDWalletProvider(
          mnemonic, `https://ropsten.infura.io/v3/${projectId}`
        )
      },
      networkId: 3,
      gasPrice: 10e10
    }
  },
};


utils/owner.js
const regeneratorRuntime = require('regenerator-runtime');
const LedgerWalletProvider = require('truffle-ledger-provider');
const HDWalletProvider = require('@truffle/hdwallet-provider');
const { setupLoader } = require('@openzeppelin/contract-loader');
const proxyAbi = JSON.parse(require('./proxyAdminAbi.json').result);
const assert = require('assert');

function requireEnvVar(envVar, error) {
  assert(process.env[envVar] !== undefined, error);
  return process.env[envVar];
}

// Logging
function log() {
  console.log.apply(this, arguments)
}

const networks = {
  rinkeby: 4,
  mainnet: 1
};

const mainnet = 'mainnet';
const rinkeby = 'rinkeby';

function getHDWallet(mnemonic, projectId, network) {
  return new HDWalletProvider(mnemonic, `https://${network}.infura.io/v3/${projectId}`);
}

function getLedgerWallet(accountOffset, projectId, network) {
  const networkId = networks[network];
  assert(!isNaN(networkId), `Invalid network ${network}`);
  assert(!isNaN(accountOffset), `Invalid accountOffset ${accountOffset}`);
  const ledgerOptions = {
    networkId,
    path: `44'/60'/${accountOffset}'/0/0`, // path: "44'/60'/7'/0/0" ledger derivation path for eth account 8 (I had to copy this from the advanced logs ledgerlive) (0xc77b79A3859D19f640de049d8E38eF573Bd9654f)
    // path: "44'/60'/8'/0/0", // ledger derivation path for eth account 9 (I had to copy this from the advanced logs ledgerlive) (0x32C8A6fba0CBb59FC9C612C862b9026e9C9Ec9D0)
    askConfirm: true,
    accountsLength: 1,
    accountsOffset: 0
  };
  return new LedgerWalletProvider(ledgerOptions, `https://${network}.infura.io/v3/${projectId}`, true);
}

async function main(provider, proxyAbi, proxyAdminAddress, newOwner) {
  const loader = setupLoader({ provider }).web3;
  const proxyAdmin = await loader.fromABI(proxyAbi, null, proxyAdminAddress);
  let currentOwner = await proxyAdmin.methods.owner().call();
  console.log(`Current owner is ${currentOwner}`);
  const res = await proxyAdmin.methods.transferOwnership(newOwner).send({ from: currentOwner, gas: 500000, gasPrice: 128e9 });
  currentOwner = await proxyAdmin.methods.owner().call();
  console.log(`Current owner is ${currentOwner}`);
}


function getMnemonic() {
  return requireEnvVar('MNEMONIC', 'HD Wallet mnemonic for deployment');
}

function getSettings(useMainnet, hdWalletToLedger) {
  let provider, proxyAdminAddress, newOwner;

  if(useMainnet) {
    const network = mainnet;
    const projectId = requireEnvVar('PROJECT_ID', 'Infura project id');
    proxyAdminAddress = '0xbFe922C94bCE67bdb0CECb9a7eB80133edD27AE8';
    if(hdWalletToLedger) {
      newOwner = '0xa7B4e8d6bAD83Bb04660171eE154c6E50c62Bf50';  // ledger top address
      provider = getHDWallet(getMnemonic(), projectId, network);
    } else {
      const accountOffset = 0;
      newOwner = '0xBD5d6b9029C5bBeE703C9277C6c6eA4dB9bd2EB4';  // hd wallet top address
      provider = getLedgerWallet(accountOffset, projectId, network);
    }
  } else {
    const network = rinkeby;
    require('dotenv').config({path: '.env.rinkeby'});
    const projectId = requireEnvVar('PROJECT_ID', 'Infura project id');
    proxyAdminAddress = '0x319EeC7bdAd9EbAC65DAd006aCBE39d5b9832ef6';
    if(hdWalletToLedger) {
      newOwner = '0x4f277c4aD936d4460D0644f08d1E968704d93625';
      provider = getHDWallet(getMnemonic(), projectId, network);
    } else {
      const accountOffset = 7;
      newOwner = '0x4f277c4aD936d4460D0644f08d1E968704d93625';
      provider = getLedgerWallet(accountOffset, projectId, network);
    }
  }

  return {provider, proxyAdminAddress, newOwner}
}

const [useMainnet, hdWalletToLedger] = [false, true];
// const [useMainnet, hdWalletToLedger] = [false, false];
// const [useMainnet, hdWalletToLedger] = [true, true];
// const [useMainnet, hdWalletToLedger] = [true, false];
const {provider, proxyAdminAddress, newOwner} = getSettings(useMainnet, hdWalletToLedger);

// Running
if (require.main === module) {
  main(provider, proxyAbi, proxyAdminAddress, newOwner);
}

truffle-config.js
const HDWalletProvider = require('@truffle/hdwallet-provider');
const assert = require('assert');
// require('dotenv').config({path: '.env.mainnet'});

function requireEnvVar(envVar, description) {
  assert(process.env[envVar] !== undefined, `Env var ${envVar} must be present${description ? ' - ' + description : ''}`);
  return process.env[envVar];
}

module.exports = {
  networks: {
    // development: {
    //   protocol: 'http',
    //   host: process.env['ETHEREUM_HOST'] || 'localhost',
    //   port: 9545,
    //   gas: 5000000,
    //   gasPrice: 5e9,
    //   network_id: '*',
    // },
    rinkeby: {
      provider: () => {
        require('dotenv').config({path: '.env.rinkeby'});
        const projectId = requireEnvVar('PROJECT_ID', 'Infura project id');
        const mnemonic = requireEnvVar('MNEMONIC', 'HD Wallet mnemonic for deployment');
        return new HDWalletProvider(
          mnemonic, `https://rinkeby.infura.io/v3/${projectId}`
        )
      },
      network_id: 4,
      gasPrice: 10e9,
      skipDryRun: true
    },
    ropsten: {
      provider: () => {
        require('dotenv').config({path: '.env.ropsten'});
        const projectId = requireEnvVar('PROJECT_ID', 'Infura project id');
        const mnemonic = requireEnvVar('MNEMONIC', 'HD Wallet mnemonic for deployment');
        return new HDWalletProvider(
          mnemonic, `https://ropsten.infura.io/v3/${projectId}`
        )
      },
      network_id: 3,
      gasPrice: 10e9,
      skipDryRun: true
    },
    mainnet: {
      provider: () => {
        const projectId = requireEnvVar('PROJECT_ID', 'Infura project id');
        const mnemonic = requireEnvVar('MNEMONIC', 'HD Wallet mnemonic for deployment');
        return new HDWalletProvider(
          mnemonic, `https://mainnet.infura.io/v3/${projectId}`
        )
      },
      network_id: 1,
      gasPrice: 42e9,
      skipDryRun: true,
      gas: 10000000
    },
  },

  // Set default mocha options here, use special reporters etc.
  mocha: {
    // timeout: 100000
  },

  api_keys: {
    etherscan: process.env['ETHERSCAN_API_KEY']
  },

  plugins: ['solidity-coverage', 'truffle-plugin-verify'],

  compilers: {
    solc: {
      version: "0.8.14",
      docker: false,
      settings: {
       optimizer: {
         enabled: false,
         runs: 200
       },
      },
    },
  },
};


utils/sendFunds.js
const HDWalletProvider = require('@truffle/hdwallet-provider');
const Web3 = require('web3');
const assert = require('assert');

function requireEnvVar(envVar, error) {
  assert(process.env[envVar] !== undefined, error);
  return process.env[envVar];
}

// Logging
function log() {
  console.log.apply(this, arguments)
}

const mainnet = 'mainnet';
const rinkeby = 'rinkeby';

function getHDWallet(mnemonic, projectId, network) {
  return new HDWalletProvider(mnemonic, `https://${network}.infura.io/v3/${projectId}`);
}

async function main(provider, amount, to) {
  try {
    const web3 = new Web3(provider);
    const accounts = await web3.eth.getAccounts();
    const amountToSend = web3.utils.toWei(amount, "ether"); //convert to wei value
    console.log('sending', accounts[0], amountToSend)
    let send = await web3.eth.sendTransaction({from: accounts[0], to, value:amountToSend, gas: 500000, gasPrice: 128e9 });
    console.log(send);
  }
  catch(e) {
    console.log(e);
  }
}


function getMnemonic() {
  return requireEnvVar('MNEMONIC', 'HD Wallet mnemonic for deployment');
}

function getSettings() {
  const network = mainnet;
  const projectId = requireEnvVar('PROJECT_ID', 'Infura project id');
  const provider = getHDWallet(getMnemonic(), projectId, network);

  return provider;
}

const provider = getSettings();

// Running
if (require.main === module) {
  main(provider, '3', '0xa7B4e8d6bAD83Bb04660171eE154c6E50c62Bf50');
}